src/io/sb3/toSb3.ts

```diff
@@ -1,163 +1,1120 @@
+  json: string;
+}
+
+export default function toSb3(options: Partial<ToSb3Options> = {}): ToSb3Output {
+  // Serialize a project. Returns an object containing the text to be stored
+  // in the caller's project.json output file. toSb3 should be bound or applied
+  // so that 'this' refers to the Project object to be serialized.
+
+  let warn = (message: string) => undefined;
+  if (options.warn) {
+    warn = options.warn;
+  }
+
+  function getVariableId(variableName: string, target: Target, stage: Stage): string | null {
+    // Get the ID associated with the provided variable name on either the
+    // provided target or the stage, or null if no matching variable is found.
+
+    let variable = target.getVariable(variableName);
+    variable = variable || stage.getVariable(variableName);
+    return variable ? variable.id : null;
+  }
+
+  function getListId(listName: string, target: Target, stage: Stage): string | null {
+    // Get the ID associated with the provided list name on either the provided
+    // target or the stage, or null if no matching list is found.
+
+    let list = target.getList(listName);
+    list = list || stage.getList(listName);
+    return list ? list.id : null;
+  }
+
+  interface SerializeInputsToFieldsOptions {
+    stage: Stage;
+    target: Target;
+
+    fieldEntries;
+  }
+
+  function serializeInputsToFields(
+    inputs: { [key: string]: BlockInput.Any },
+    options: SerializeInputsToFieldsOptions
+  ): {
+    fields: { [key: string]: any };
+  } {
+    // Serialize provided inputs into a "fields" mapping that can be stored
+    // on a serialized block.
+    //
+    // Where the Scratch 3.0 term "input" refers to a slot that accepts blocks,
+    // the term "field" is any user-interactive slot that cannot be obscured as
+    // such. Fields are also the interactive element within any shadow block,
+    // making their presence key to nearly all inputs.
+    //
+    // While inputs are fairly complex to serialize, fields are comparatively
+    // simple. A field always contains only its selected or entered value, so
+    // it's not concerned with a variety of structures like an input is.
+    // The format for a field mapping looks something like this:
+    //
+    //   {
+    //     VARIABLE: ["my variable", "theVariablesId"]
+    //   }
+    //
+    // Some fields take an ID; some don't. Here's another example:
+    //
+    //   {
+    //     DISTANCETOMENU: ["_mouse_"]
+    //   }
+    //
+    // The important thing to remember with fields during serialization is that
+    // they refer to slots that don't accept blocks: non-droppable menus,
+    // primarily, but the value in a (non-compressed) shadow input too.
 
-import List from "../../List";
-import Variable from "../../Variable";
+    const { fieldEntries, stage, target } = options;
 
-export default async function toSb3(project: Project): Promise<ArrayBuffer> {
-  const outZip = new JSZip();
+    const result = {
+      fields: {}
+    };
```

Comment:
Same here re: structuring/destructuring roundtrip when only one value is returned